### PR TITLE
core: audit.meta.__internalOptionalArtifacts

### PR DESCRIPTION
**Pull request by connorjclark**
optionalArtifacts allow an audit to request an artifact that may enhance the analysis, but isn't necessary to get useful information. Intended for artifacts that are experimental (SourceMaps), or expensive shouldn't be required (also, maybe SourceMaps).

Changed the terminology inside `config.js` to use `requestedArtifacts`. The warning messages were already using it, so it fits well :)

see also: #10090

### CODE REVIEW

File: lighthouse-core/runner.js [hunk 1]
```diff
@@ -328,17 +328,19 @@ class Runner {
...
+      // Only pass the declared `requiredArtifacts`/`optionalArtifacts` to the audit
       // The type is masquerading as `LH.Artifacts` but will only contain a subset of the keys
       // to prevent consumers from unnecessary type assertions.
-      const requiredArtifacts = audit.meta.requiredArtifacts
-        .reduce((requiredArtifacts, artifactName) => {
+      const requestedArtifacts = audit.meta.requiredArtifacts
+        .concat(audit.meta.optionalArtifacts || []);
+      const narrowedArtifacts = requestedArtifacts
+        .reduce((narrowedArtifacts, artifactName) => {
           const requiredArtifact = artifacts[artifactName];
```
Comment (patrickhulce) [inline lighthouse-core/runner.js]: while we're incurring all the rename churn anyway... :)
```suggestion
          const requestedArtifact = artifacts[artifactName];
```

File: lighthouse-core/test/config/config-test.js [hunk 2]
```diff
@@ -147,6 +147,74 @@ describe('Config', () => {
...
+        gatherers: [
+          'viewport-dimensions',
+        ],
+      }],
+      audits: [DoesntNeedYourCrap],
+    });
+    expect(config.passes[0].gatherers.map(g => g.path)).toEqual(['viewport-dimensions']);
+  });
+
+  it('does warn when an audit requests and receives an optional artifact', async () => {
```
Comment (patrickhulce) [inline lighthouse-core/test/config/config-test.js]:
```suggestion
  it('should keep optionalArtifacts in gatherers after filter', async () => { 
```

or something? the warn is just copy/paste cruft right

File: lighthouse-core/test/config/config-test.js [hunk 3]
```diff
@@ -147,6 +147,74 @@ describe('Config', () => {
...
+      passes: [{
+        passName: 'defaultPass',
+        gatherers: [
+          'source-maps',
+          'viewport-dimensions',
+        ],
+      }],
+      audits: [ButWillStillTakeYourCrap],
+    });
+    expect(config.passes[0].gatherers.map(g => g.path))
```
Comment (patrickhulce) [inline lighthouse-core/test/config/config-test.js]: isn't this going to pass without your change? I think we need to add `onlyAudits: ['optional-artifact-audit']` to trigger filtering

File: lighthouse-core/config/config.js [hunk 4]
```diff
@@ -69,7 +69,7 @@ function assertValidPasses(passes, audits) {
     pass.gatherers.forEach(gathererDefn => {
       const gatherer = gathererDefn.instance;
       foundGatherers.add(gatherer.name);
-      const isGatherRequiredByAudits = requiredGatherers.has(gatherer.name);
```
Comment (patrickhulce) [inline lighthouse-core/config/config.js]: should we warn in the scenario that an optional artifact is missing? I think that depends on our imagined long-term use case and how close to "required" these optionalArtifacts are supposed to be

**Comment by connorjclark** [comment]
yeah, I suppose something like `--skip-gatherers` or a more powerful config extending system (https://github.com/GoogleChrome/lighthouse/issues/7406 touches on this) would be needed to really utilize this.

I'm ok with the private rename.

File: lighthouse-core/test/config/config-test.js (hunk 2 above)
Comment (connorjclark) [reply lighthouse-core/test/config/config-test.js]: no just brain fart :)

File: lighthouse-core/config/config.js (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/config/config.js]: I don't like warnings that are "maybe this is a warning, IDK, you figure it out!" :)

File: lighthouse-core/test/config/config-test.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/test/config/config-test.js]: yea. consider this a regression test.

good point on triggering filtering.

Comment (connorjclark) [reply lighthouse-core/test/config/config-test.js]: also need to have categories to filter ... so I just changed it to extend the default config.

File: lighthouse-core/config/config.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/config/config.js]: well in the case where we're trying to pull it in for them but can't that seems like prime warning territory it's basically just a notch below required and a notch below complete failure is a warning.

if it's the scenario where they're truly optional and we don't try to pull the gatherer in for them, then I agree sounds like no warning is appropriate.

IMO it's more egregious at the moment that the default CLI run results in warnings for unused artifacts that are only there for potential plugin usage.

Comment (patrickhulce) [reply lighthouse-core/config/config.js]: since it sounds like we're going with the former, but internal only, I'm fine with no warning.

**Comment by patrickhulce** [comment]
> I'm ok with the private rename.

let's do it then if no one else has feelings :) we can always upgrade to expose it to everyone in a minor release if it becomes useful

File: types/audit.d.ts [hunk 5]
```diff
@@ -55,6 +55,8 @@ declare global {
       description: string;
       /** A list of the members of LH.Artifacts that must be present for the audit to execute. */
       requiredArtifacts: Array<keyof Artifacts>;
+      /** A list of the members of LH.Artifacts that augment the audit, but aren't necessary. */
```
Comment (patrickhulce) [inline types/audit.d.ts]:
```suggestion
      /** A list of the members of LH.Artifacts that augment the audit, but aren't necessary. For internal use only with experimental-config. */
```

**Pull request merged**